File: native/sensor-bridge/android/sensor-bridge/src/main/kotlin/com/dragonfly/sensorbridge/GlucoseEvent.kt
package com.dragonfly.sensorbridge

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Mirrors NewGlucoseReading in packages/shared/src/types.ts. */
@Serializable
data class GlucoseEvent(
    val valueMgDl: Double,
    val timestamp: String, // ISO-8601 UTC
    val context: Context,
    val ingestionPath: IngestionPath,
    val readingKind: ReadingKind = ReadingKind.SENSOR,
    val trend: Trend? = null,
    val rawDeviceId: String? = null,
    val vendor: Vendor? = null,
    val deviceName: String? = null,
    val notes: String? = null,
) {
    @Serializable
    enum class Vendor {
        @SerialName("dexcom") DEXCOM,
        @SerialName("libre") LIBRE,
        @SerialName("unknown") UNKNOWN,
    }

    @Serializable
    enum class Trend {
        @SerialName("rising_quickly") RISING_QUICKLY,
        @SerialName("rising") RISING,
        @SerialName("rising_slowly") RISING_SLOWLY,
        @SerialName("flat") FLAT,
        @SerialName("falling_slowly") FALLING_SLOWLY,
        @SerialName("falling") FALLING,
        @SerialName("falling_quickly") FALLING_QUICKLY,
        @SerialName("unknown") UNKNOWN,
    }

    @Serializable
    enum class ReadingKind {
        @SerialName("sensor") SENSOR,
        @SerialName("backfill") BACKFILL,
        @SerialName("manual") MANUAL,
    }

    @Serializable
    enum class IngestionPath {
        @SerialName("native-ble") NATIVE_BLE,
        @SerialName("native-nfc") NATIVE_NFC,
        @SerialName("healthkit") HEALTHKIT,
        @SerialName("health-connect") HEALTH_CONNECT,
        @SerialName("manual") MANUAL,
    }

    @Serializable
    enum class Context {
        @SerialName("pre_taiyi") PRE_TAIYI,
        @SerialName("post_taiyi") POST_TAIYI,
        @SerialName("before_lunch") BEFORE_LUNCH,
        @SerialName("post_lunch_1_to_2h") POST_LUNCH_1_TO_2H,
        @SerialName("post_lunch_3_to_4h") POST_LUNCH_3_TO_4H,
        @SerialName("end_of_day") END_OF_DAY,
    }
}

/** Convert mmol/L to mg/dL (rounded). */
fun mmolPerLToMgDl(mmol: Double): Double = kotlin.math.round(mmol * 18.0182)

File: native/sensor-bridge/ios/DragonflySensorBridge/Sources/DragonflySensorBridge/GlucoseEvent.swift
import Foundation

/// Mirrors `NewGlucoseReading` in `packages/shared/src/types.ts`.
/// Adapter implementations must produce values that can be JSON-encoded
/// directly into the sync batch sent to `POST /api/glucose/sync`.
public struct GlucoseEvent: Codable, Equatable {
    public enum Vendor: String, Codable, Equatable, CaseIterable {
        case dexcom, libre, unknown
    }
    public enum Trend: String, Codable, Equatable, CaseIterable {
        case risingQuickly = "rising_quickly"
        case rising
        case risingSlowly = "rising_slowly"
        case flat
        case fallingSlowly = "falling_slowly"
        case falling
        case fallingQuickly = "falling_quickly"
        case unknown
    }
    public enum ReadingKind: String, Codable, Equatable, CaseIterable {
        case sensor, backfill, manual
    }
    public enum IngestionPath: String, Codable, Equatable, CaseIterable {
        case nativeBle = "native-ble"
        case nativeNfc = "native-nfc"
        case healthKit = "healthkit"
        case healthConnect = "health-connect"
        case manual
    }
    public enum Context: String, Codable, Equatable, CaseIterable {
        case preTaiyi = "pre_taiyi"
        case postTaiyi = "post_taiyi"
        case beforeLunch = "before_lunch"
        case postLunch1To2h = "post_lunch_1_to_2h"
        case postLunch3To4h = "post_lunch_3_to_4h"
        case endOfDay = "end_of_day"
    }

    public var valueMgDl: Double
    public var timestamp: Date
    public var context: Context
    public var trend: Trend?
    public var rawDeviceId: String?
    public var readingKind: ReadingKind
    public var ingestionPath: IngestionPath
    public var vendor: Vendor?
    public var deviceName: String?
    public var notes: String?

    public init(
        valueMgDl: Double,
        timestamp: Date,
        context: Context,
        trend: Trend? = nil,
        rawDeviceId: String? = nil,
        readingKind: ReadingKind = .sensor,
        ingestionPath: IngestionPath,
        vendor: Vendor? = nil,
        deviceName: String? = nil,
        notes: String? = nil
    ) {
        self.valueMgDl = valueMgDl
        self.timestamp = timestamp
        self.context = context
        self.trend = trend
        self.rawDeviceId = rawDeviceId
        self.readingKind = readingKind
        self.ingestionPath = ingestionPath
        self.vendor = vendor
        self.deviceName = deviceName
        self.notes = notes
    }
}

/// Convert mmol/L to mg/dL. Many EU Libre devices report mmol/L.
public func mmolPerLToMgDl(_ mmol: Double) -> Double {
    return (mmol * 18.0182).rounded()
}
